lib/spack/spack/package.py

```diff
@@ -689,6 +691,23 @@ def global_license_file(self):
         return os.path.join(self.global_license_dir, self.name,
                             os.path.basename(self.license_files[0]))
 
+    @property
+    def unique_id(self):
+        """An identifier with the package name prepended for readability."""
+        if not self.spec.concrete:
+            raise ValueError("Cannot provide a unique, readable id when "
+                             "the spec is not concretized.")
+        # TODO: Restore use of the dag hash once resolve issues with different
+        # TODO: hashes being associated with dependents of different packages
+        # TODO: within the same install, such as the hash for callpath being
+        # TODO: different for mpich and dyninst in the
+        # TODO: test_force_uninstall_and_reinstall_by_hash` test.
```

Comment:
I'm confused about how the DAG hash can actually conflict for single-root DAGs.  That shouldn't be possible.  Under one root you can only have one spec with a given name, so there shouldn't be multiple versions of anything.

Reply:
Revisited this problem so I can provide details.  (This is easy to reproduce by including the hash in `unique_id`.)

Build tasks are created with a post-order traversal of the spec.  In this case there are the following entries when re-installing `callpath-1.0-lbzbz32w5aiiesugnacmcht2frcksb4v`:

- `libelf-0.8.13-hve5eehmhg4cuyyo7cgc6eujyea7443j`
- `mpich-3.0.4-4v3nsfvxc2xufhgrzphzwfgeonkji5d6`
- `libdwarf-20130729-wvuwyhovcnduu6sbjzaptxgotjsmwjb5`
- `dyninst-8.2-xvb3xy52riui7jyceawrf3v3p4k5m4qk`
- `callpath-1.0-lbzbz32w5aiiesugnacmcht2frcksb4v`  (dependent of `mpich`)

Which means the entry for the following is "missing" when including the hash in the id (or key):
-  `callpath-1.0-nwc2p47aq2tmt4w2ab5pqnrwqqztjo76` (dependent of `dyninst`)

When`dyninst` is installed, it tries to remove itself from the uninstalled dependencies of `callpath-1.0-nwc2p47...` (but can't).   

Then, when the priority of `callpath-1.0-lbzbz32...` is checked prior to attempting the installation, it fails because `dyninst` was not removed from its uninstalled dependencies.

So that's what I meant by a "conflict".